### PR TITLE
[MAINTENANCE] Refactor `ValidationAction` to use Pydantic

### PR DESCRIPTION
**Pull request by cdkini**
Since our new V1 checkpoints utilize Pydantic and want to support a rich-type instantiation pattern, we should make any nested params also leverage Pydantic.
```python
checkpoint = Checkpoint(
    validations=[ValidationDefinition(...)],
    actions=[SlackAction(...)],
)
```
We currently utilize dictionary configs so there is some complexity in removing the old `instantiate_class_from_config` pattern. Noted that nested within actions are renderers, which have a similar issue; these need to eventually be migrated to Pydantic as well (deeming this out of scope as well).

- [x] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [x] Code is linted - run `invoke lint` (uses `ruff format` + `ruff check`)
- [x] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: great_expectations/checkpoint/actions.py [hunk 1]
```diff
@@ -58,6 +63,7 @@ class ValidationAction:
     def __init__(self) -> None:
         from great_expectations import project_manager
 
+        super().__init__()
```
Comment (cdkini) [inline great_expectations/checkpoint/actions.py]: How can I introduce project manager injection without overriding `__init__`?

File: great_expectations/checkpoint/actions.py [hunk 2]
```diff
@@ -128,6 +134,21 @@ class (for open source Great Expectations) or the GeCloudIdentifier (from Great
...
+            runtime_environment={},
+            config_defaults={},
+        )
+        if not renderer:
+            raise ClassInstantiationError(
+                module_name=config.get("module_name"),
+                package_name=None,
+                class_name=config.get("class_name"),
+            )
+        return renderer
```
Comment (cdkini) [inline great_expectations/checkpoint/actions.py]: I'd like to accept the `Renderer` rich type but that introduces more complexity (need to update that class hierarchy to also use Pydantic). For now, let's continue to accept a dictionary config but build the rich type just-in-time when it needs to be used.

File: great_expectations/checkpoint/actions.py [hunk 3]
```diff
@@ -179,42 +200,35 @@ class SlackNotificationAction(DataDocsAction):
...
+    slack_webhook: Optional[str] = None
+    slack_token: Optional[str] = None
+    slack_channel: Optional[str] = None
+    notify_on: Literal["all", "failure", "success"] = "all"
+    notify_with: Optional[list[str]] = None
+    show_failed_expectations: bool = False
+    renderer: dict = {
+        "class_name": "SlackRenderer",
+        "module_name": "great_expectations.render.renderer.slack_renderer",
+    }
```
Comment (cdkini) [inline great_expectations/checkpoint/actions.py]: Note that I've added default values for renderers. I'd imagine this covers 99% of usecases

File: great_expectations/checkpoint/actions.py [hunk 4]
```diff
@@ -179,42 +200,35 @@ class SlackNotificationAction(DataDocsAction):
...
+
+        try:
+            if not slack_token and slack_channel:
+                assert slack_webhook
+            if not slack_webhook:
+                assert slack_token and slack_channel
+            assert not (slack_webhook and slack_channel and slack_token)
+        except AssertionError:
+            raise ValueError(
+                "Please provide either slack_webhook or slack_token and slack_channel"
```
Comment (cdkini) [inline great_expectations/checkpoint/actions.py]: Shifted any validation from `__init__` to validators

File: great_expectations/checkpoint/actions.py (hunk 2 above)
Comment (cdkini) [reply great_expectations/checkpoint/actions.py]: I'm providing a very sane default for all renderers so the chance a user is going to override it is very slim. We can follow up with cleanup to accept rich type renderers (but I don't think we should prioritize it)

File: great_expectations/checkpoint/actions.py [hunk 5]
```diff
@@ -128,12 +160,30 @@ class (for open source Great Expectations) or the GeCloudIdentifier (from Great
...
+    _build_data_docs: Callable = PrivateAttr()
+    _get_docs_sites_urls: Callable = PrivateAttr()
+
+    def __init__(self, **data: Any) -> None:
         from great_expectations import project_manager
 
-        super().__init__()
+        super().__init__(**data)
         self._build_data_docs = project_manager.build_data_docs
         self._get_docs_sites_urls = project_manager.get_docs_sites_urls
```
Comment (cdkini) [inline great_expectations/checkpoint/actions.py]: Is this the pattern for Pydantic private attrs? Feels a bit bad to override init?

File: great_expectations/checkpoint/actions.py [hunk 6]
```diff
@@ -165,11 +215,9 @@ class SlackNotificationAction(DataDocsAction):
...
     Args:
         renderer: Specifies the Renderer used to generate a query consumable by Slack API, e.g.:
-            ```python
-            {
-               "module_name": "great_expectations.render.renderer.slack_renderer",
-               "class_name": "SlackRenderer",
-           }
+           ```python
+            "module_name": "great_expectations.render.renderer.slack_renderer",
+            "class_name": "SlackRenderer",
```
Comment (cdkini) [inline great_expectations/checkpoint/actions.py]: Netlify issues and I don't know why 😠

File: great_expectations/checkpoint/actions.py [hunk 7]
```diff
@@ -179,42 +227,41 @@ class SlackNotificationAction(DataDocsAction):
...
+    renderer: dict = {
+        "class_name": "SlackRenderer",
+        "module_name": "great_expectations.render.renderer.slack_renderer",
+    }
+
+    _renderer: SlackRenderer = PrivateAttr()
+
+    def __init__(self, **data: Any) -> None:
+        super().__init__(**data)
+        self._renderer = cast(SlackRenderer, self._build_renderer(config=self.renderer))
```
Comment (cdkini) [inline great_expectations/checkpoint/actions.py]: Need to cast for all of these - is there a better solution?

File: docs/docusaurus/docs/snippets/checkpoints_and_actions.py [hunk 8]
```diff
@@ -438,7 +438,6 @@
       class_name: SlackNotificationAction
       slack_webhook: <YOUR SLACK WEBHOOK URL>
       notify_on: failure
-      notify_with: all
```
Comment (cdkini) [inline docs/docusaurus/docs/snippets/checkpoints_and_actions.py:441]: Should be an empty list if we want all - simplified the constructor below

File: tests/actions/test_core_actions.py [hunk 9]
```diff
@@ -648,3 +671,121 @@ def test_cloud_sns_notification_action(
...
+        ),
+    ],
+)
+def test_action_serialization(
+    mock_context,
+    action_class: Type[ValidationAction],
+    init_params: dict,
+    expected: dict,
+):
+    action = action_class(**init_params)
```
Comment (cdkini) [inline tests/actions/test_core_actions.py]: Need to instantiate the action within the test since we need the mock context setup first

File: great_expectations/checkpoint/actions.py (hunk 5 above)
Comment (cdkini) [reply great_expectations/checkpoint/actions.py]: Note this also ruins autocomplete in editors

File: tests/actions/test_core_actions.py [hunk 10]
```diff
@@ -648,3 +671,138 @@ def test_cloud_sns_notification_action(
...
+            MicrosoftTeamsNotificationAction,
+            OpsgenieAlertAction,
+            EmailAction,
+            UpdateDataDocsAction,
+            SNSNotificationAction,
+            APINotificationAction,
+        ],
+        actual,
+    )
+    assert isinstance(parsed_obj, action_class)
```
Comment (cdkini) [inline tests/actions/test_core_actions.py]: Trying to keep this as simple as possible but maybe not the best approach

File: great_expectations/checkpoint/actions.py (hunk 5 above)
Comment (cdkini) [reply great_expectations/checkpoint/actions.py]: Both of these classes are parent classes and won't be exposed to the end user during normal workflows

File: great_expectations/checkpoint/actions.py [hunk 11]
```diff
@@ -179,42 +236,44 @@ class SlackNotificationAction(DataDocsAction):
...
-        )
-        module_name = renderer["module_name"]
-        if not self.renderer:
-            raise ClassInstantiationError(
-                module_name=module_name,
-                package_name=None,
-                class_name=renderer["class_name"],
+    class Config:
+        arbitrary_types_allowed = True
+        json_encoders = {SlackRenderer: lambda s: _encode_renderer(s)}
```
Comment (cdkini) [inline great_expectations/checkpoint/actions.py]: Decided to use rich types in the `__init__` but write class_name/module_name to disk when serializing

File: great_expectations/checkpoint/actions.py [hunk 12]
```diff
@@ -8,12 +8,41 @@
...
+)
+from great_expectations.render.renderer.microsoft_teams_renderer import (
+    MicrosoftTeamsRenderer,
+)
+from great_expectations.render.renderer.opsgenie_renderer import (
+    OpsgenieRenderer,
+)
+from great_expectations.render.renderer.slack_renderer import (
+    SlackRenderer,
+)
```
Comment (billdirks) [inline great_expectations/checkpoint/actions.py]: Can you move these all to `great_expectations/render/renderer/__init__.py` so we can import them from 1 place?

File: great_expectations/checkpoint/actions.py (hunk 6 above)
Comment (billdirks) [reply great_expectations/checkpoint/actions.py]: You mean the reason you had to remove `{` and `}`? Do you have a link to the error?

File: great_expectations/checkpoint/actions.py (hunk 5 above)
Comment (billdirks) [reply great_expectations/checkpoint/actions.py]: What do you mean it breaks autocomplete? 

Pydantic private attrs mean you have to manage it yourself so you need to explicitly set them. 
You could make them properties and then you wouldn't have to override `__init__()`. Would this let you avoid breaking autocomplete?

File: great_expectations/checkpoint/actions.py [hunk 13]
```diff
@@ -130,10 +187,13 @@ class (for open source Great Expectations) or the GeCloudIdentifier (from Great
 
 
 class DataDocsAction(ValidationAction):
-    def __init__(self) -> None:
+    _build_data_docs: Callable = PrivateAttr()
+    _get_docs_sites_urls: Callable = PrivateAttr()
+
+    def __init__(self, **data: Any) -> None:
```
Comment (billdirks) [inline great_expectations/checkpoint/actions.py]: What is data here and why do you need it? Is it really an `Any`? It looks like it is defined in `ValidationAction` so all subclasses should have it in their `__init__` methods so this was previously a Liskov violation.

File: great_expectations/checkpoint/actions.py [hunk 14]
```diff
@@ -179,42 +237,45 @@ class SlackNotificationAction(DataDocsAction):
...
+    slack_channel: Optional[str] = None
+    notify_on: Literal["all", "failure", "success"] = "all"
+    notify_with: Optional[List[str]] = None
+    show_failed_expectations: bool = False
+    renderer: SlackRenderer = SlackRenderer()
+
+    @validator("renderer", pre=True)
+    def _validate_renderer(cls, renderer: dict | SlackRenderer) -> SlackRenderer:
+        if isinstance(renderer, dict):
+            renderer = cast(SlackRenderer, _build_renderer(config=renderer))
```
Comment (billdirks) [inline great_expectations/checkpoint/actions.py]: Can you validate the cast?

```
if renderer.__class__ != "SlackRenderer":
    raise ...
```

File: great_expectations/checkpoint/actions.py [hunk 15]
```diff
@@ -179,42 +237,45 @@ class SlackNotificationAction(DataDocsAction):
...
+        return renderer
+
+    @root_validator
+    def _root_validate_slack_params(cls, values: dict) -> dict:
+        slack_token = values["slack_token"]
+        slack_channel = values["slack_channel"]
+        slack_webhook = values["slack_webhook"]
+
+        try:
+            if not slack_token and slack_channel:
```
Comment (billdirks) [inline great_expectations/checkpoint/actions.py]: I think this needs to be `if not (slack_token and slack_channel)` since:
```
>>> not True and False
False
>>> not (True and False)
True
```

Comment (billdirks) [reply great_expectations/checkpoint/actions.py]: I think all these `if` and `asserts` can be simplified to:
```
if webhook:
  assert not slack_token and not slack_channel
else:
  assert slack_token and slack_channel
```

File: great_expectations/checkpoint/actions.py [hunk 16]
```diff
@@ -349,22 +410,18 @@ class PagerdutyAlertAction(ValidationAction):
...
-        """Create a PagerdutyAlertAction"""
+    type: Literal["pagerduty"] = "pagerduty"
+
+    api_key: str
+    routing_key: str
+    notify_on: Literal["all", "failure", "success"] = "failure"
+    severity: Literal["critical", "error", "warning", "info"] = "critical"
+
+    @root_validator
+    def _root_validate_deps(cls, values: dict) -> dict:
```
Comment (billdirks) [inline great_expectations/checkpoint/actions.py:418]: I would update the pypd error message to say something like `"ModuleNotFoundError: No module named 'pypd'". "pypd" is required for PageDuty notification actions.`

I would also consider moving the `import` to here since I don't like the pattern of setting imports to `None` when they don't exist. This is the place we do this check.

File: great_expectations/checkpoint/actions.py [hunk 17]
```diff
@@ -457,30 +514,24 @@ class MicrosoftTeamsNotificationAction(ValidationAction):
...
+    teams_webhook: str
+    notify_on: Literal["all", "failure", "success"] = "all"
+    renderer: MicrosoftTeamsRenderer = MicrosoftTeamsRenderer()
+
+    @validator("renderer", pre=True)
+    def _validate_renderer(
+        cls, renderer: dict | MicrosoftTeamsRenderer
+    ) -> MicrosoftTeamsRenderer:
+        if isinstance(renderer, dict):
+            renderer = cast(MicrosoftTeamsRenderer, _build_renderer(config=renderer))
```
Comment (billdirks) [inline great_expectations/checkpoint/actions.py]: Similar comment here and below about validating the cast.

File: great_expectations/checkpoint/actions.py [hunk 18]
```diff
@@ -845,16 +876,25 @@ class StoreValidationResultAction(ValidationAction):
         TypeError: validation_result_id must be of type ValidationResultIdentifier or GeCloudIdentifier, not {}.
     """
 
+    type: Literal["store_validation_result"] = "store_validation_result"
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    _target_store: Store = PrivateAttr()
+
```
Comment (billdirks) [inline great_expectations/checkpoint/actions.py:887]: This action still takes a `data_context` as an `__init__` arg. It seems like it should take a `data` argument. This should this be updated. It seems it was like this before and can be updated in a separate PR. If not as a fast follow PR to this one, it should be ticketed and added to this sprint.

File: tests/actions/test_core_actions.py [hunk 19]
```diff
@@ -648,3 +671,138 @@ def test_cloud_sns_notification_action(
...
+):
+    action = action_class(**init_params)
+
+    # Test serialization
+    json_dict = action.json()
+    actual = json.loads(json_dict)
+    assert actual == expected
+
+    # Test deserialization
+    parsed_obj = parse_obj_as(
```
Comment (billdirks) [inline tests/actions/test_core_actions.py]: This tests seems to verify it is possible to parse these json strings. But it is not invoking gx code to do it. It seems like we should wrap this `parse_obj_as` call into a gx function and then call it here. Later when we write the code to deserialize it, we will have this function. This `Union` is ugly but maybe ok for now though it should be refactored so we don't have to manually maintain a `Union` if possible.

File: tests/actions/test_core_actions.py (hunk 10 above)
Comment (billdirks) [reply tests/actions/test_core_actions.py]: It might be nice to also go through the `init_params`, do a `getattr` on the `parsed_obj` using the key and making sure the returned value matches.

**Review by billdirks — COMMENTED** [review]
Thanks for doing this. I am out tomorrow so you might need to ping someone else to review.

File: great_expectations/checkpoint/actions.py (hunk 13 above)
Comment (cdkini) [reply great_expectations/checkpoint/actions.py]: In order to set the private attr, we override `BaseModel.__init__`. This will be any params from any of the subclasses so it needs to be a `**kwargs`.

I'd rather we not do this but I'm not sure how to set private attrs without some kind of post init behavior. You can't set with a root validator and we don't have an explicit post init hook like we would with dataclasses.

File: great_expectations/checkpoint/actions.py (hunk 18 above)
Comment (cdkini) [reply great_expectations/checkpoint/actions.py:887]: I'll be deleting this action once we get validation definitions to store their results. These changes are just to ensure tests pass for now

File: great_expectations/checkpoint/actions.py [hunk 20]
```diff
@@ -130,12 +172,27 @@ class (for open source Great Expectations) or the GeCloudIdentifier (from Great
 
 
 class DataDocsAction(ValidationAction):
-    def __init__(self) -> None:
+    def _build_data_docs(
```
Comment (tyler-hoffman) [inline great_expectations/checkpoint/actions.py:174]: Is this called?

File: great_expectations/checkpoint/actions.py (hunk 6 above)
Comment (tyler-hoffman) [reply great_expectations/checkpoint/actions.py]: To render curly brackets, you need to escape them, since they have meaning in markdown.

File: great_expectations/checkpoint/actions.py [hunk 21]
```diff
@@ -915,7 +976,7 @@ def _basic_run(
         if self._using_cloud_context and expectation_suite_identifier:
             expectation_suite_id = expectation_suite_identifier.id
 
-        return self.target_store.set(
+        return self._target_store.set(  # type: ignore[func-returns-value]
```
Comment (tyler-hoffman) [inline great_expectations/checkpoint/actions.py]: fyi I'm changing the return type of `set` in a PR I'm working on.

**Review by tyler-hoffman — APPROVED** [review]
Looks reasonable to me

File: great_expectations/checkpoint/actions.py (hunk 20 above)
Comment (cdkini) [reply great_expectations/checkpoint/actions.py:174]: Yup L1063 in `UpdateDataDocs` action
